mod-source-record-manager-server/src/main/java/org/folio/services/flowcontrol/RawRecordsFlowControlServiceImpl.java

```diff
@@ -0,0 +1,149 @@
+   * When current state equals or below di.flow.control.records.threshold - all raw records consumers from consumer's group should resume.
+   */
+  private final AtomicInteger currentState = new AtomicInteger(0);
+
+  @PostConstruct
+  public void init() {
+    LOGGER.info("Flow control feature is {}", enableFlowControl ? "enabled" : "disabled");
+  }
+
+  /**
+   * This method schedules resetting state of flow control. By default it triggered each 5 mins.
+   * This is distributed system and count of raw records pushed to process can be not corresponding
+   * with DI_COMPLETE/DI_ERROR, for example because of imports are stuck.
+   *
+   * For example, batch size - 10, max simultaneous - 20, records threshold - 10.
+   * Flow control receives 20 records and pause consumers, 5 DI_COMPLETE events came, 15 was missed due to some DB issue.
+   * Current state would equals to 15 and threshold will never met 10 records so consumers will be paused forever
+   * and as a result all subsequent imports will not have a chance to execute.
+   *
+   * This scheduled method to reset state intended to prevent that. The correlation between max simultaneous and records
+   * threshold can be missed during resetting that can cause that resume/pause cycle may not be as usual, because we are
+   * starting from clear state after reset, but any events would not be missed and consumers never pause forever.
+   */
+  @Scheduled(cron = "${di.flow.control.reset.state.cron:0 0/5 * * * ?}")
+  public void resetState() {
+    currentState.getAndUpdate(prev -> {
+      if (prev != 0) {
+        LOGGER.info("State has been reset to initial value, current value: 0");
+        return 0;
+      } else {
+        return prev;
+      }
+    });
+    resumeIfThresholdAllows();
+  }
+
+  @Override
+  public void trackChunkReceivedEvent(Integer recordsCount) {
+    if (!enableFlowControl) {
+      return;
+    }
+
+    int current = currentState.addAndGet(recordsCount);
+
+    LOGGER.info("--------------- Current value after chunk received: {} ---------------", current);
+
+    if (current >= maxSimultaneousRecords) {
+      Collection<KafkaConsumerWrapper<String, String>> rawRecordsReadConsumers = consumersStorage.getConsumersByEvent(DI_RAW_RECORDS_CHUNK_READ.value());
+
+      rawRecordsReadConsumers.forEach(consumer -> {
+        if (consumer.demand() > 0) {
+          consumer.pause();
+
+          LOGGER.info("Kafka consumer - id: {}, subscription - {} is paused, because {} exceeded {} max simultaneous records",
+            consumer.getId(), DI_RAW_RECORDS_CHUNK_READ.value(), current, maxSimultaneousRecords);
+        }
+      });
+    }
+  }
+
+  @Override
+  public void trackChunkDuplicateEvent(Integer recordsCount) {
+    if (!enableFlowControl) {
+      return;
+    }
+
+    int prev = currentState.get();
+
+    currentState.set(prev - recordsCount);
```

Comment:
i think getAndUpdate should be used to perform the previous two lines atomically.